design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+When a resource fits under this category, the name should be a verb, using camel-case if necessary.
+
+### Resource Representation
+
+The standard format for resource representation is JSON. There is no rule for schemas or the semantics used by a resource representation, as long as they follow the JSON syntax properly.
+
+However, it is important to taken into account the following principles when defining resource representations:
+
+* Keep it simple for clients, so that a representation can be easily built and processed when making requests/responses to the API
+* Be aware of extensive representations as they can affect overall performance and usability, if possible review your API to see if resources can be decomposed into smaller ones
+* Avoid deeply nested JSON objects
+* Avoid using generic resource representations
+
+### Http Status Codes
+
+The tables below defines the HTTP status code that clients should expect from APIs.
+
+This document also defines which codes are expected to be used for each HTTP verb, please take a look at [HTTP Verbs](#httpverbs) for more details.
+
+Clients should consider the status codes herein defined in order to properly handle responses from APIs, recover from errors, and 
+understand the reasoning behind the API responses.
+
+#### Success Codes
+
+|Verb|Description|
+|---|---|
+|200|Indicates the request completed successfully regardless the size of the response body|
+|201|Indicates the request completed successfully and a resource was created|
+|204|Indicates the request completed successfully and the server did not provide a response body|
+
+#### Redirection Codes
+
+|Verb|Description|
+|---|---|
+|307|Indicates that clients should resubmit the request to another location|
+
+#### Client Side Error Codes
+
+|Verb|Description|
+|---|---|
+|400|Indicates that the request is invalid, usually related with the validation of the payload|
+|401|Indicates that clients should provide authorization or the provided authorization is invalid|
+|403|Indicates that the authorization provided by the client is not enough to access the resource|
+|404|Indicates that the requested resource does not exist|
+|405|Indicates that the method chosen by the client to access a resource is not supported|
+|409|Indicates that the resource the client is trying to create already exists or some conflict when processing the request|
+
+#### Server Side Error Codes
+
+|Verb|Description|
+|---|---|
+|500|Indicates that the server could not fulfill the request due to some unexpected error|
+
+### Http Verbs
+
+Based on the REST architectural style, HTTP Verbs should be used as the means to communicate to APIs the operations to be performed
+on their resources. While HTTP verbs cover by themselves the most common operations you might want to look the [Controller Resource](#controllerresource) as an alternative on how to overcome any limitation imposed by the usage of HTTP verbs herein defined.
```

Comment:
The "Controller Resource" link is broken.

Reply:
@vmuzikar marking as resolved, because it was already updated. Thank you.